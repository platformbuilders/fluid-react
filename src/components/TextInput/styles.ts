import styled from 'styled-components/native';
import { Animated } from 'react-native';
import { ifStyle, switchStyle, getTheme } from '../../utils/helpers';
import { TextInput as TextInputType, InputStatus } from '../../utils/types';

interface InputAreaWrapperProps {
  multiline: boolean;
}

interface BottomLineProps {
  dark?: boolean;
  status?: string;
}

const isMultiline = ifStyle('multiline');
const switchStatus = switchStyle('status');
const primaryContrast = getTheme('primary.contrast');
const primaryDark = getTheme('primary.dark');
const success = getTheme('success');
const failure = getTheme('failure');
const disabled = getTheme('disabled');
const inputColor = (props: BottomLineProps): any =>
  switchStatus({
    [InputStatus.SUCCESS]: success,
    [InputStatus.FAILURE]: failure,
    [InputStatus.DEFAULT]: props.dark ? primaryDark : primaryContrast,
    [InputStatus.DISABLED]: disabled,
  });

export const LABEL_UPPER_STYLE = {
  top: 8,
  fontSize: 14,
};

export const LABEL_LOWER_STYLE = {
  top: 40,
  fontSize: 18,
};

interface WrapperProps {
  multiline: boolean;
}

export const Wrapper = styled.View<WrapperProps>`
  height: ${isMultiline('auto', '89px')};
  max-height: ${isMultiline('auto', '89px')};
  min-height: ${isMultiline('89px', 'auto')};
  padding-top: 8px;
  position: relative;
  justify-content: flex-end;
`;

export const TextLabel = styled.Text<TextInputType>`
  line-height: 19px;
  position: absolute;
  color: ${inputColor};
  font-size: ${LABEL_LOWER_STYLE.fontSize}px;
`;

export const Label = Animated.createAnimatedComponent(TextLabel);

export const InputAreaWrapper = styled.View<InputAreaWrapperProps>`
  margin-top: 6px;
  margin-bottom: 9px;
  flex-direction: row;
  min-height: 24px;
  max-height: ${isMultiline('auto', '24px')};
`;

export const TextInput = styled.TextInput<TextInputType>`
  padding: 0;
  flex-grow: 1;
  border-width: 0;
  min-height: 24px;
  margin-top: ${isMultiline('16px', '0px')};
  max-height: ${isMultiline('150px', '24px')};
  font-weight: 700;
  color: ${inputColor};
`;

export const BottomLine = styled.View<BottomLineProps>`
  height: 1px;
  background-color: ${inputColor};
`;
