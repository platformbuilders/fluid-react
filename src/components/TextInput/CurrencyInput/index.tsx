import React, { FC } from 'react';
import MaskedInput from 'react-text-mask';
import createNumberMask from 'text-mask-addons/dist/createNumberMask';
import { FormError } from '../..';
import { toOnlyNumbers } from '../../../utils/helpers';

import { Bar, Input, Label, Wrapper } from './styles';

type Props = {
  id: string;
  name?: string;
  label?: string;
  value: string | number | string[] | undefined;
  error?: string | boolean;
  onChangeText?: (value: any) => void;
};

const defaultMaskOptions = {
  prefix: 'R$ ',
  suffix: '',
  includeThousandsSeparator: true,
  thousandsSeparatorSymbol: '.',
  allowDecimal: true,
  decimalSymbol: ',',
  decimalLimit: 2,
};

const CurrencyInputComponent: FC<Props> = ({
  error,
  label,
  id,
  name,
  onChangeText,
  value,
}) => {
  const currencyMask = createNumberMask(defaultMaskOptions);

  const handleChange = (event: any) => {
    if (onChangeText && event.target.value) {
      onChangeText(toOnlyNumbers(event.target.value));
    }
  };

  return (
    <FormError error={error}>
      <Wrapper>
        <MaskedInput
          mask={currencyMask}
          placeholder="R$ 0,00"
          id={id}
          name={name}
          value={value}
          onChange={handleChange}
          render={(ref, props) => <Input ref={ref} error={error} {...props} />}
        />
        <Bar className="bar" error={error} />
        <Label error={error}>{label}</Label>
      </Wrapper>
    </FormError>
  );
};

export default CurrencyInputComponent;
