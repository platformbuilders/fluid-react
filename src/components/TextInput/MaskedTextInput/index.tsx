import React, { FC, useContext } from 'react';
import { ThemeProvider } from 'styled-components';
import { MaskedTextInput as MaskedTextInputType } from '~/utils/types';
import { TextInput } from './styles';
import { ThemeContext } from '../../ThemeContext';

const MaskedTextInput: FC<MaskedTextInputType> = ({
  maskType,
  dark = false,
  multiline = false,
  ...props
}) => {
  const { theme } = useContext(ThemeContext);
  return (
    <ThemeProvider theme={theme}>
      <TextInput
        {...props}
        dark={dark}
        multiline={multiline}
        type={maskType}
        underlineColorAndroid="transparent"
      />
    </ThemeProvider>
  );
};

export default MaskedTextInput;
