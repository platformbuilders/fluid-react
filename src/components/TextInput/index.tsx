import { ChangeEvent, FC, RefObject, useRef } from 'react';
import { IMaskMixin } from 'react-imask';

import { TextInputType } from '../../types';
import { Input, Label, Message, PlaceholderLabel, Wrapper } from './styles';

const InputMask = IMaskMixin(({ inputRef, ...props }) => (
  <Input {...props} ref={inputRef as RefObject<HTMLInputElement> | undefined} />
));

const TextInput: FC<TextInputType> = ({
  message,
  onChange,
  onBlur,
  onFocus,
  style,
  label,
  textInputStyle,
  value,
  id,
  name,
  maxLength,
  autoFocus,
  maskOptions,
}) => {
  const hasValue = value?.length > 0;

  const onAccept = (_, __, event?: InputEvent | undefined) => {
    onChange(event as Partial<ChangeEvent<any>>);
  };

  const ref = useRef(null);
  const inputRef = useRef(null);

  return (
    <Wrapper style={style}>
      <Label>
        {maskOptions ? (
          <InputMask
            {...maskOptions}
            id={id}
            name={name}
            data-testid={id}
            style={textInputStyle}
            ref={ref}
            inputRef={inputRef}
            onAccept={onAccept}
            autoFocus={autoFocus}
          />
        ) : (
          <Input
            id={id}
            data-testid={id}
            name={name}
            value={value}
            style={textInputStyle}
            onBlur={onBlur}
            onFocus={onFocus}
            onChange={onChange}
            maxLength={maxLength}
            autoFocus={autoFocus}
          />
        )}
        <PlaceholderLabel $hasValue={hasValue}>{label}</PlaceholderLabel>
        {message ? <Message>{message}</Message> : null}
      </Label>
    </Wrapper>
  );
};

export default TextInput;
