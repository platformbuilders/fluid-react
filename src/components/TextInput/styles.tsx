import React from 'react';
import styled from 'styled-components';
import { StylesProvider } from '@material-ui/styles';
import { TextField } from '@material-ui/core';
import { getTheme } from '../../utils/helpers';

const primaryMain = getTheme('primary.main');

export const Input = styled((props) => (
  <StylesProvider injectFirst>
    <TextField {...props} />
  </StylesProvider>
))`
  border-color: red;
  & label {
    font-size: 1em;
  }
  & label.Mui-focused {
    color: ${primaryMain};
  }
  & .MuiInput-underline:after {
    border-bottom-color: ${primaryMain};
  }
`;

export const InputWrapper = styled.div`
  flex-direction: column;
`;
