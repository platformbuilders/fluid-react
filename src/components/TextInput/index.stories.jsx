import React from 'react';
import { boolean, text, withKnobs } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import markdown from './textInput.md';
import TextInput from './index';

storiesOf('TextInput', module)
  .addDecorator(withKnobs)
  .addParameters({
    readme: {
      content: markdown,
    },
    viewport: {
      defaultViewport: '',
    },
  })
  .add('Example', () => (
    <TextInput
      label={text('label', 'Name')}
      name={text('name', 'name')}
      placeholder={text('placeholder', 'Your Name')}
      fullWidth={boolean('fullWidth', false)}
      autoFocus={boolean('autoFocus', false)}
      maskType="cep"
      variant="filled"
    />
  ));
