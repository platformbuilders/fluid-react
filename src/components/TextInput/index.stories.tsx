import type { Meta, StoryObj } from '@storybook/react';
import TextInput from './index';

const meta: Meta<typeof TextInput> = {
  title: 'TextInput',
  component: TextInput,
  args: {
    label: 'Name',
  },
};

export default meta;
type Story = StoryObj<typeof TextInput>;

export const Default: Story = {};
export const Mask: Story = {
  args: {
    label: 'Cep',
    maskOptions: { mask: '00000-000' },
  },
};
