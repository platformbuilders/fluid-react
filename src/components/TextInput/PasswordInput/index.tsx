import React, { useState, FC, useCallback, useContext } from 'react';
import { ThemeProvider } from 'styled-components';
import { TextInputProps } from '../types';
import { TextInput } from '~/components';
import { ThemeContext } from '../../ThemeContext';

const PasswordInput: FC<TextInputProps> = (props) => {
  const [hidePassword, setHidePassword] = useState<boolean>(true);
  const { theme } = useContext(ThemeContext);
  const hitSlop = {
    left: 40,
    right: 40,
    top: 40,
    bottom: 40,
  };

  const onPressShowPassword = useCallback((): void => {
    setHidePassword(!hidePassword);
  }, [hidePassword]);

  return (
    <ThemeProvider theme={theme}>
      <TextInput
        secureTextEntry={hidePassword}
        iconName={hidePassword ? 'eye' : 'eye-off'}
        iconTouchableEnabled
        onPressIcon={onPressShowPassword}
        iconHitSlop={hitSlop}
        {...props}
      />
    </ThemeProvider>
  );
};

export default PasswordInput;
