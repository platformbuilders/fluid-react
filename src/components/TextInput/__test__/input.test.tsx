import { ThemeProvider } from 'styled-components';
import { fireEvent, render, screen } from '@testing-library/react';

import theme from '../../../theme';
import { TextInputType } from '../../../types';
import TextInput from '../index';

const onChange = jest.fn();

const inputId = 'input-test';

const defaultProps: TextInputType = {
  id: inputId,
  value: '',
  label: 'label',
  name: 'inputTest',
  onChange,
};

describe('Component: TextInput', () => {
  beforeEach(() => {
    jest.restoreAllMocks();
    jest.clearAllMocks();
    jest.resetModules();
  });

  test('snapshots with default props', () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        <TextInput {...defaultProps} />
      </ThemeProvider>,
    );
    expect(container).toMatchSnapshot();
  });

  test('should change value of input without mask to 456', () => {
    render(<TextInput {...defaultProps} />);

    const inputElement: any = screen.getByTestId(inputId);

    fireEvent.change(inputElement, { target: { value: 456 } });

    expect(onChange).toHaveBeenCalledTimes(1);
  });

  test('should change value of input with mask to 123', () => {
    render(<TextInput {...defaultProps} maskOptions={{ mask: Number }} />);

    const inputElement: any = screen.getByTestId(inputId);

    fireEvent.change(inputElement, { target: { value: 123 } });

    expect(inputElement.value).toBe('123');
  });
});
