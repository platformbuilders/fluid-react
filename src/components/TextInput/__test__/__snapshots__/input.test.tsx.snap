// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: TextInput snapshots with default props 1`] = `
<div>
  <div
    class="sc-eCYdKt hSGTba"
  >
    <label
      class="sc-hKMtDE"
    >
      <input
        class="sc-gsnScm lmkUye"
        data-testid="input-test"
        id="input-test"
        name="inputTest"
        value=""
      />
      <span
        class="sc-bczSft cNBLmY"
      >
        label
      </span>
    </label>
  </div>
</div>
`;
